Port SharedBuffer to the new IPC serialization format

cdumez opened the pull request:
#### 2b82b0b362a5b7b4b1bac7437577834635e392bb
<pre>
Port SharedBuffer to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=268669">https://bugs.webkit.org/show_bug.cgi?id=268669</a>

Reviewed by Alex Christensen.

* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::SharedBuffer::create):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::SharedBuffer::asFragmentedSharedBuffer const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;WebCore::SharedBuffer&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;WebCore::SharedBuffer&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/274046@main">https://commits.webkit.org/274046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f93d49cd3e393cb20960c6c9799ba0e6521a466a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37630 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16510 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39878 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40161 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33500 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38947 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13671 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31879 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38195 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13899 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32973 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12135 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12079 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33656 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41422 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33995 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34080 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37981 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12663 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10197 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36150 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14099 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13063 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4889 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13412 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -299,6 +301,8 @@ class SharedBuffer : public FragmentedSharedBuffer {
     WEBCORE_EXPORT GRefPtr<GBytes> createGBytes() const;
 #endif
 
+    Ref<FragmentedSharedBuffer> asFragmentedSharedBuffer() const { return const_cast<SharedBuffer&>(*this); }
```
Can't this return a Ref<const FragmentedSharedBuffer> rather than introduce a const_cast?

cdumez replied:
What do you mean by returning a Ref? Doesn’t it already return a Ref? I tried several things but haven’t been able to get rid of this const cast. I’d be happy to try more things though.

cdumez replied:
If you meant `Ref<const WebCore::FragmentedSharedBuffer> asFragmentedSharedBuffer();`, I tried but the IPC decoder does like it:
```
/Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/ArgumentCoder.h:61:15: error: calling a protected constructor of class 'WebCore::FragmentedSharedBuffer'
            T t;
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/GeneratedSerializers.mm:27:
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/GeneratedWebKitSecureCoding.h:28:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:28:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/Shared/Cocoa/CoreIPCArray.h:30:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:28:
/Volumes/Work/WebKit/OpenSource/Source/WebKit/Platform/IPC/ArgumentCoders.h:309:34: note: in instantiation of function template specialization 'IPC::ArgumentCoder<const WebCore::FragmentedSharedBuffer>::decode<IPC::Decoder>' requested here
        return ArgumentCoder<U>::decode(decoder);
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/GeneratedSerializers.mm:27:
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/GeneratedWebKitSecureCoding.h:28:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:28:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/Shared/Cocoa/CoreIPCArray.h:30:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:28:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebKit/Platform/IPC/ArgumentCoders.h:30:
/Volumes/Work/WebKit/OpenSource/Source/WebKit/Platform/IPC/Decoder.h:159:40: note: in instantiation of function template specialization 'IPC::ArgumentCoder<WTF::Ref<const WebCore::FragmentedSharedBuffer>>::decode<IPC::Decoder, const WebCore::FragmentedSharedBuffer>' requested here
            std::optional<T> t { Impl::decode(*this) };
/Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/GeneratedSerializers.mm:45001:45: note: in instantiation of function template specialization 'IPC::Decoder::decode<WTF::Ref<const WebCore::FragmentedSharedBuffer>>' requested here
    auto asFragmentedSharedBuffer = decoder.decode<Ref<const WebCore::FragmentedSharedBuffer>>();
```

cdumez committed:
Port SharedBuffer to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=268669

Reviewed by Alex Christensen.

* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::SharedBuffer::create):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::SharedBuffer::asFragmentedSharedBuffer const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::SharedBuffer>::encode): Deleted.
(IPC::ArgumentCoder<WebCore::SharedBuffer>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/274046@main
